components/tooltip/script.js

```diff
@@ -65,91 +78,114 @@ class Tooltip extends HTMLElement {
-            this.setPosition(this.getVisiblePosition(overflows));
-        }
+        return [orientation, elementPosition];
     }
 
-    overflows() {
-        var rect = this.getBoundingClientRect();
-        const overflows = {};
-
-        if (rect.top < 0) overflows.top = true;
-        if (rect.left < 0)  overflows.left = true;
-        if (rect.right > (window.innerWidth || document.documentElement.clientWidth)) overflows.right = true;
-        if (rect.bottom > (window.innerHeight || document.documentElement.clientHeight)) overflows.bottom = true;
+    /**
+     * Sets the new position of the tooltip. Waits 2 frames for the tooltip to be fully rendered
+     * and checks if the tooltip is visible on the new position. If it is not - calls setPosition
+     * recursively until a suitable position is found or we've ran out of possibilities.
+     * @param {number} scrollOffsetX - the offset of the scroll on the X axis.
+     * @param {number} scrollOffsetY - the offset of the scroll on the Y axis.
+     * @param {string} orientation - the current position of the tooltip - top, left, right or bottom.
+     * @returns {promise}
+    */
+    async setPosition(scrollOffsetX, scrollOffsetY, orientation = this.position) {
+        const elementSize = this.triggerElement.getBoundingClientRect();
 
-        return overflows;
-    }
+        let [updatedOrientation, elementPosition] = this.getPositionCoords(orientation, elementSize);
 
-    getVisiblePosition(overflows) {
-        const allSides = ['top', 'bottom', 'left', 'right'];
+        this.position = updatedOrientation;
+        this.style.top = scrollOffsetY + elementPosition.top + 'px';
+        this.style.left = scrollOffsetX + elementPosition.left + 'px';
 
-        overflows = overflows || this.overflows();
+        await this.waitForFrames(2);
+        const overflows = this.overflows;
         let overflowingSides = Object.keys(overflows);
+        if (overflowingSides.length && this.uncheckedOrientations.length !== 0) {
+            return await this.setPosition(scrollOffsetX, scrollOffsetY, this.getVisiblePosition(overflowingSides));
+        }
 
-        overflowingSides = overflowingSides.filter(side => this.fixedSides.indexOf(side) == -1);
+        return Promise.resolve();
+    }
 
-        let nextIndex = allSides.indexOf(this.position) + 1;
-        nextIndex = [(nextIndex + allSides.length) % allSides.length];
-        this.position = allSides[nextIndex];
-        this.fixedSides.push(this.position);
+    /**
+     * An asynchronous method that waits for given amount of frames before it resolves;
+     * useful in cases where we need to wait for styles to be computed.
+     * @param {number} count - the number of frames it should wait for before resolving.
+     * @returns {promise}
+     */
+    async waitForFrames(count) {
+        while(count--) {
+            await new Promise((resolve) => requestAnimationFrame(resolve));
+        }
+        return Promise.resolve();

```

Comment:
You can skip and remove `return Promise.resolve()` because it is async method and it will be resolved automatically.

Reply:
Mhmm, removed in both usages.